src/java/org/apache/cassandra/index/sai/iterators/KeyRangeIntersectionIterator.java

```diff
@@ -77,7 +77,19 @@ protected PrimaryKey computeNext()
                 if (index != alreadyAvanced)
                 {
                     KeyRangeIterator range = ranges.get(index);
-                    PrimaryKey nextKey = nextOrNull(range, highestKey);
+                    PrimaryKey nextKey = range.getCurrent();
+
+                    // Note that we will either have a data model that produces SKINNY primary keys or a data model
+                    // that produces some combination of WIDE and STATIC prikary keys.
+                    if (nextKey.kind() == PrimaryKey.Kind.WIDE || nextKey.kind() == highestKey.kind())
+                        // We can always skip if the target is of the same kind or this range is non-static. 
+                        nextKey = nextOrNull(range, highestKey);
+                    else if (nextKey.kind() == PrimaryKey.Kind.STATIC && nextKey.compareTo(highestKey) < 0)
+                        // For a range of static keys, only skip if we'e advanced to a new partition, and when we
+                        // do, skip to an actual static key. We may otherwise skip too far, as static row IDs always
+                        // precede non-static ones in on-disk postings lists.
+                        nextKey = nextOrNull(range, highestKey.toStatic());
+
                     if (nextKey == null || nextKey.compareTo(highestKey) > 0)
```

Comment:
What if `highestKey` is STATIC? It may mess up the algorithm if there are other `WIDE` keys, because it would basically make this condition false, as STATIC key == any wide keys in the same partition. So it could break intersecting other wide keys within the partition.

I think this would manifest if we had 2 regular columns and intersect it with a static column.
Fortunately the intersection clause limit = 2 would not allow it, but a user can change it and allow more than 2 columns in an intersection.

Reply:
Would it break end-to-end, or would it just return a `STATIC` `PrimaryKey` from the intersection, which would potentially cause more post-filtering than we want in `ResultRetriever`? In any case, I think in the case where we have a `STATIC` `highestKey`, and `WIDE` `PrimaryKey` can be treated as higher, and we could probably make that change without changing any of the existing `compareTo()` behavior, which seems to open up a can of worms.

In short, I think we could fix this doing something like...

```
if (nextKey == null || nextKey.compareTo(highestKey) > 0 || (nextKey.compareTo(highestKey) == 0 && nextKey.kind() == WIDE && highestKey.kind() == STATIC))
```

...or just add a `compareToStrict()` to `PrimaryKey` that encapsulates this. We **DO** want to avoid a case where we post-filter a whole partition when 2 `WIDE` key iterators don't match anything from each other in the context of a match from a `STATIC` key iterator. Let me work on this a bit...

Reply:
I've added a commit and further testing that should address this. Let me know what you think.

Reply:
`compareToStrict` looks good. Also thanks for adding a test that exposes that problem.